src/reactionsystem.jl

```diff
@@ -694,76 +712,115 @@ function ReactionSystem(rxs::Vector, iv = Catalyst.DEFAULT_IV; kwargs...)
-                                      spatial_ivs = nothing, kwargs...)
+# Called internally (whether DSL-based or programmtic model creation is used). 
+# Creates a sorted reactions + equations vector, also ensuring reaction is first in this vector.
+# Extracts potential species, variables, and parameters from the input (if not provided as part of 
+# the model creation) and creates the corresponding vectors. 
+# While species are ordered before variables in the unknowns vector, this ordering is not imposed here,
+# but carried out at a later stage.
+function make_ReactionSystem_internal(rxs_and_eqs::Vector, iv, us_in, ps_in; spatial_ivs = nothing, 
+                                      continuous_events = [], discrete_events = [], kwargs...)
+
+    # Creates a combined iv vector (iv and sivs). This is used later in the function (so that 
+    # independent variables can be exluded when encountered quantities are added to `us` and `ps`).
     t = value(iv)
     ivs = Set([t])
     if (spatial_ivs !== nothing)
         for siv in (MT.scalarize(spatial_ivs))
             push!(ivs, value(siv))
         end
     end
-    sts = OrderedSet{eltype(sts_in)}(sts_in)
+
+    # Initialises the new unknowns and parameter vectors.
+    # Preallocates the `vars` set, which is used by `findvars!`
+    us = OrderedSet{eltype(us_in)}(us_in)
     ps = OrderedSet{eltype(ps_in)}(ps_in)
     vars = OrderedSet()
 
+    # Extracts the reactions and equations from the combined reactions + equations input vector.
     all(eq -> eq isa Union{Reaction, Equation}, rxs_and_eqs)
     rxs = Reaction[eq for eq in rxs_and_eqs if eq isa Reaction]
     eqs = Equation[eq for eq in rxs_and_eqs if eq isa Equation]
 
-    # add species / parameters that are substrates / products first
-    for rx in rxs, reactants in (rx.substrates, rx.products)
-        for spec in reactants
-            MT.isparameter(spec) ? push!(ps, spec) : push!(sts, spec)
-        end
-    end
-
+    # Loops through all reactions, adding encountered quantities to the unknown and parameter vectors.
     for rx in rxs
-        findvars!(ps, sts, rx.rate, ivs, vars)
-        for s in rx.substoich
-            (s isa Symbolic) && findvars!(ps, sts, s, ivs, vars)
+        # Loops through all reaction substrates and products, extracting these.
+        for reactants in (rx.substrates, rx.products), spec in reactants
+            MT.isparameter(spec) ? push!(ps, spec) : push!(us, spec)
         end
-        for p in rx.prodstoich
-            (p isa Symbolic) && findvars!(ps, sts, p, ivs, vars)
+
+        # Adds all quantitites encountered in the reaction's rate.
+        findvars!(ps, us, rx.rate, ivs, vars)
+
+        # Extracts all quantitites encountered within stoichiometries.
+        for stoichiometry in (rx.substoich, rx.prodstoich), sym in stoichiometry
+            (sym isa Symbolic) && findvars!(ps, us, sym, ivs, vars)
         end
```

Comment:
Aren't these changes going to alter the order of parameters and species potentially? That may break tests for downstream packages (which really shouldn't be relying on order, but I think some may).

Reply:
Yes it will.

Long term this felt like a better way to phrase the code, especially as we add more steps (like checking various metadata fields).

As you say, stuff should not rely on order, but I agree it might break _something_. That said, if it improves the code I don't think we should adapt for someone doing something they shouldn't do. In either case we def should note this in the History file.

I am still OK with reverting this though.

Reply:
As long are you are ok with making updates to downstream unmaintained packages we'd like to keep working that break from this I am ok with the change. (i.e. stuff like SBMLToolkit, FiniteStateProjection, etc.) I just don't want to have to go and submit PRs to these various packages if they rely on order and we change it now.